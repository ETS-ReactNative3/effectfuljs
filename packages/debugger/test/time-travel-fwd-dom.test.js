require("./setup-time-travel-fwd");
const trace = require("./run");
trace.silent = true;
const D = require("../main");
const SD = require("@effectful/serialization/dom");
const S = require("@effectful/serialization");

test("DOM forward time traveling", function() {
  D.TimeTravel.reset();
  global.console = { log: jest.fn() };
  D.TimeTravel.reset();
  SD.trackGlobalDocument();
  document.body.innerHTML = `<div id="root"></div>`;
  const mod = trace(D.evalThunk(require("./__fixtures__/dom")));
  console.log("> M:", mod);
  D.TimeTravel.DOM.track(document.getElementById("root"));
  const text0 = document.body.innerHTML;
  expect(text0).toMatchSnapshot();
  D.TimeTravel.checkpoint();
  trace(mod.changeDom1());
  const text1 = document.body.innerHTML;
  expect(text1).toMatchSnapshot();
  D.TimeTravel.checkpoint();
  trace(mod.changeDom2());
  const text2 = document.body.innerHTML;
  expect(text2).toMatchSnapshot();
  D.TimeTravel.checkpoint();
  trace(mod.changeDom3());
  const text3 = document.body.innerHTML;
  expect(text3).toMatchSnapshot();
  D.TimeTravel.checkpoint();
  trace(mod.changeDom4());
  const text4 = document.body.innerHTML;
  expect(text4).toMatchSnapshot();
  D.TimeTravel.undo();
  expect(document.body.innerHTML).toBe(text3);
  D.TimeTravel.redo();
  expect(document.body.innerHTML).toBe(text4);
  D.TimeTravel.undo();
  expect(document.body.innerHTML).toBe(text3);
  D.TimeTravel.undo();
  expect(document.body.innerHTML).toBe(text2);
  D.TimeTravel.redo();
  expect(document.body.innerHTML).toBe(text3);
  D.TimeTravel.redo();
  expect(document.body.innerHTML).toBe(text4);
  D.TimeTravel.undo();
  expect(document.body.innerHTML).toBe(text3);
  D.TimeTravel.undo();
  expect(document.body.innerHTML).toBe(text2);
  D.TimeTravel.undo();
  expect(document.body.innerHTML).toBe(text1);
  D.TimeTravel.undo();
  expect(document.body.innerHTML).toBe(text0);
  D.TimeTravel.redo();
  expect(document.body.innerHTML).toBe(text1);
  D.TimeTravel.redo();
  expect(document.body.innerHTML).toBe(text2);
  D.TimeTravel.redo();
  expect(document.body.innerHTML).toBe(text3);
  D.TimeTravel.redo();
  expect(document.body.innerHTML).toBe(text4);
  D.TimeTravel.redo();
  expect(document.body.innerHTML).toBe(text4);
  D.TimeTravel.undo();
  expect(document.body.innerHTML).toBe(text3);
  const j = D.TimeTravel.journal;
  const data = S.write({
    now: j.now,
    future: j.future,
    past: j.past,
    data: document.body
  });
  expect(JSON.stringify(data)).toMatchSnapshot();
  const oldDoc = document.documentElement;
  D.TimeTravel.DOM.untrack(document.getElementById("root"));
  document.body.innerHTML = "";
  D.TimeTravel.reset();
  const restored = S.read(data);
  document.body.replaceWith(restored.data);
  D.TimeTravel.DOM.track(document.getElementById("root"));
  expect(document.body.innerHTML).toBe(text3);
  Object.assign(D.TimeTravel.journal, restored);
  D.TimeTravel.redo();
  expect(document.body.innerHTML).toBe(text4);
  D.TimeTravel.redo();
  expect(document.body.innerHTML).toBe(text4);
  D.TimeTravel.undo();
  expect(document.body.innerHTML).toBe(text3);
  D.TimeTravel.undo();
  expect(document.body.innerHTML).toBe(text2);
  D.TimeTravel.undo();
  expect(document.body.innerHTML).toBe(text1);
  D.TimeTravel.undo();
  expect(document.body.innerHTML).toBe(text0);
  D.TimeTravel.undo();
  expect(document.body.innerHTML).toBe(text0);
  D.TimeTravel.redo();
  expect(document.body.innerHTML).toBe(text1);
  D.TimeTravel.redo();
  expect(document.body.innerHTML).toBe(text2);
  D.TimeTravel.redo();
  expect(document.body.innerHTML).toBe(text3);
  D.TimeTravel.redo();
  expect(document.body.innerHTML).toBe(text4);
  D.TimeTravel.redo();
  expect(document.body.innerHTML).toBe(text4);
  expect(console.log.mock.calls).toMatchSnapshot();
});
