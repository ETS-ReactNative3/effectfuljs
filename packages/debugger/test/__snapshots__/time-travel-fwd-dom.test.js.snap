// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DOM forward time traveling 1`] = `"<div id=\\"root\\"></div>"`;

exports[`DOM forward time traveling 2`] = `"<div id=\\"root\\"><div style=\\"color:green\\" id=\\"i1\\">Hi</div></div>"`;

exports[`DOM forward time traveling 3`] = `"<div id=\\"root\\"><div style=\\"color:red\\" id=\\"i1\\"><div id=\\"i2\\">There</div></div></div>"`;

exports[`DOM forward time traveling 4`] = `"<div id=\\"root\\"><div style=\\"color:blue\\" id=\\"i1\\"><h1 id=\\"i3\\"></h1><div id=\\"i2\\">There</div><h2 id=\\"i4\\"></h2></div></div>"`;

exports[`DOM forward time traveling 5`] = `"<div id=\\"root\\"><div style=\\"color:blue\\" id=\\"i1\\"><h1 id=\\"i3\\"></h1><div id=\\"i2\\">THERE</div></div></div>"`;

exports[`DOM forward time traveling 6`] = `"{\\"f\\":[[\\"now\\",{\\"f\\":[[\\"prev\\",null],[\\"operations\\",{\\"i\\":[{\\"ty\\":\\"attributes\\",\\"el\\":{\\"r\\":4},\\"an\\":[],\\"rn\\":[],\\"tn\\":\\"style\\",\\"ov\\":\\"color:red\\",\\"$\\":\\"MutationRecord\\"},{\\"ty\\":\\"childList\\",\\"el\\":{\\"r\\":4},\\"an\\":[{\\"r\\":0}],\\"rn\\":[],\\"ns\\":{\\"r\\":2},\\"$\\":\\"MutationRecord\\"},{\\"ty\\":\\"childList\\",\\"el\\":{\\"r\\":4},\\"an\\":[{\\"r\\":3}],\\"rn\\":[],\\"ps\\":{\\"r\\":2},\\"$\\":\\"MutationRecord\\"}],\\"p\\":{\\"$\\":\\"undefined\\"},\\"$\\":\\"DomSnapshot\\"}]]}],[\\"future\\",{\\"f\\":[[\\"prev\\",null],[\\"operations\\",{\\"i\\":[{\\"ty\\":\\"childList\\",\\"el\\":{\\"r\\":4},\\"an\\":[{\\"r\\":3}],\\"rn\\":[],\\"ps\\":{\\"r\\":2},\\"$\\":\\"MutationRecord\\"},{\\"ty\\":\\"characterData\\",\\"el\\":{\\"r\\":1},\\"an\\":[],\\"rn\\":[],\\"ov\\":\\"THERE\\",\\"$\\":\\"MutationRecord\\"}],\\"p\\":{\\"$\\":\\"undefined\\"},\\"$\\":\\"DomSnapshot\\"}]]}],[\\"past\\",{\\"f\\":[[\\"prev\\",{\\"f\\":[[\\"prev\\",null],[\\"operations\\",{\\"i\\":[{\\"ty\\":\\"childList\\",\\"el\\":{\\"r\\":5},\\"an\\":[{\\"r\\":4}],\\"rn\\":[],\\"$\\":\\"MutationRecord\\"}],\\"p\\":{\\"$\\":\\"undefined\\"},\\"$\\":\\"DomSnapshot\\"}]]}],[\\"operations\\",{\\"i\\":[{\\"ty\\":\\"childList\\",\\"el\\":{\\"r\\":4},\\"an\\":[],\\"rn\\":[{\\"text\\":\\"Hi\\",\\"$\\":\\"Text\\"}],\\"$\\":\\"MutationRecord\\"},{\\"ty\\":\\"childList\\",\\"el\\":{\\"r\\":4},\\"an\\":[{\\"r\\":2}],\\"rn\\":[],\\"$\\":\\"MutationRecord\\"},{\\"ty\\":\\"attributes\\",\\"el\\":{\\"r\\":4},\\"an\\":[],\\"rn\\":[],\\"tn\\":\\"style\\",\\"ov\\":\\"color:green\\",\\"$\\":\\"MutationRecord\\"}],\\"p\\":{\\"$\\":\\"undefined\\"},\\"$\\":\\"DomSnapshot\\"}]]}],[\\"data\\",{\\"tag\\":\\"BODY\\",\\"c\\":[{\\"r\\":5}],\\"a\\":{},\\"$\\":\\"Element\\"}]],\\"x\\":[{\\"tag\\":\\"H1\\",\\"c\\":[],\\"a\\":{\\"id\\":\\"i3\\"},\\"f\\":[[\\"innerText\\",\\"PRE\\"]],\\"$\\":\\"Element\\"},{\\"text\\":\\"There\\",\\"$\\":\\"Text\\"},{\\"tag\\":\\"DIV\\",\\"c\\":[{\\"r\\":1}],\\"a\\":{\\"id\\":\\"i2\\"},\\"f\\":[[\\"innerHtml\\",\\"hi there\\"]],\\"$\\":\\"Element\\"},{\\"tag\\":\\"H2\\",\\"c\\":[],\\"a\\":{\\"id\\":\\"i4\\"},\\"f\\":[[\\"innerText\\",\\"post\\"]],\\"$\\":\\"Element\\"},{\\"tag\\":\\"DIV\\",\\"c\\":[{\\"r\\":0},{\\"r\\":2},{\\"r\\":3}],\\"a\\":{\\"style\\":\\"color:blue\\",\\"id\\":\\"i1\\"},\\"$\\":\\"Element\\"},{\\"tag\\":\\"DIV\\",\\"c\\":[{\\"r\\":4}],\\"a\\":{\\"id\\":\\"root\\"},\\"$\\":\\"Element\\"}]}"`;

exports[`DOM forward time traveling 7`] = `
Array [
  Array [
    "> M:",
    Object {
      "__esModule": true,
      "changeDom1": [Function],
      "changeDom2": [Function],
      "changeDom3": [Function],
      "changeDom4": [Function],
    },
  ],
]
`;
